### ISSUE TITLE
Extract retrieval of security key and signing credentials into their own services.

### ISSUE THREAD

**Issue by acollard** [IdentityServer/IdentityServer3#1393]
We'd like to use HMAC SHA-256 instead of X509 security key for signing. 
X509Certificate is hard coded in the following areas:
- TokenValidator
- DefaultTokenSigningService

I've replaced DefaultTokenSigningService with my custom one that supports any SigningCredentials.  But I can't easily replace the X509SecurityKey in the TokenValidator without replacing it completely.

Ideally the IOC container would support interfaces for the SecurityKey and SigningCredentials factories.

Something like this:
```c#
  public interface ISecurityKeyService
  {
    SecurityKey GetSecurityKey(string clientId);
  }

  public interface ISigningCredentialsService
  {
    SigningCredentials GetSigningCredentials(string clientId);
  }
```

By default you would have the following X509 implementation:
```c#

  public class X509SecurityKeyService : ISecurityKeyService
  {
    readonly X509Certificate2 _Cert;

    public X509SecurityKeyService(X509Certificate2 cert)
    {
      _Cert = cert;
    }

    public SecurityKey GetSecurityKey(string clientId)
    {
      return new X509SecurityKey(_Cert);
    }
  }

  public class X509SigningCredentialsService : ISigningCredentialsService
  {
    readonly X509Certificate2 _Cert;

    public X509SigningCredentialsService(X509Certificate2 cert)
    {
      _Cert = cert;
    }

    public SigningCredentials GetSigningCredentials(string clientId)
    {
      return new X509SigningCredentials(_Cert);
    }
  }
```

**Comment by leastprivilege** [comment]
For which token types do you want to do that? id or access - or both?

**Comment by acollard** [comment]
Both, I don't see any reason to have different SecurityKeys/SigningCredentials for id and access tokens. 

We don't need this but I expect you will support SecurityKeys based on which client is used. So each client could have their own SecurityKey.

We have this running at the moment (by registering our own TokenValidator and TokenSigningService) and we are just using a SharedSecurityKeyService.

```c#
  public class SharedSecurityKeyService : ISecurityKeyService
  {
    readonly SecurityKey _SecurityKey;
    public SharedSecurityKeyService(SecurityKey securityKey)
    {
      Guard.ArgumentNotNull(securityKey, "securityKey");
      _SecurityKey = securityKey;
    }

    public SecurityKey GetSecurityKey(string clientId = null)
    {
      return _SecurityKey;
    }
  }
```

**Comment by leastprivilege** [comment]
Well - the OIDC spec allows either asymmetric (e.g. x509) or symmetrically (e.g. HMACSHA256) using the client secret. We removed that feature since our client secrets are hashed and that would have forced us to to store them using reversible encryption.

For access tokens - how do you want to associate key material with the token? per client (does not make sense). Per scope? What about situations where multiple scopes are requested but they used different signing keys?

**Comment by acollard** [comment]
Sorry it's been a while since I've worked on this stuff, coming back to me now. 

To contradict my earlier point, you would likely want different keys for the Id and Access tokens. In a web app scenario you would want to use X509 to validate the id token and you could use symmetric on the access token used by the web service. 

You are right the keys would have to be tied to the scope for the access token. We should support cases where two or more scopes share the same key. But in the case that they are different I see two options:
1. Fall back to some shared key. Likely X509 keys found through the jwks_uri.
2. Throw an error. Obviously not ideal and I don't know if it works into the OIDC spec.

**Comment by leastprivilege** [comment]
And what's your reason to use symmetric instead of x509 for access tokens?

**Comment by acollard** [comment]
We were trying to shrink the size of the token. In certain scenarios we need to pass the auth tokens in the query params. Ex. retrieving images from our web service through an img tag.

We ran into IIS maxQueryString default of 2048. We've increased this for now, but I want to make sure we didn't hit this again. So firstly we are trying to reduce the amount of information we pass back in claims.

Secondly I thought we could switch to HMAC SHA-256. I know the difference between signature sizes HMAC (64 bytes, I think) and X509 (256 bytes, I think) isn't that large, but I was thinking every little bit helps. I expected HMAC to be supported based on https://openid.net/specs/draft-jones-json-web-token-07.html#Algorithms.

> Of the JWS signing algorithms, only HMAC SHA-256 MUST be implemented by conforming JWT implementations. It is RECOMMENDED that implementations also support the RSA SHA-256 and ECDSA P-256 SHA-256 algorithms.
